api/client/proto/authservice.proto

```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+// credential.
+message ChangePasswordWithTokenRequest {
+    string TokenID = 1;
+    // SecondFactorToken is an otp value.
+    string SecondFactorToken = 2;
+    // Password is user's password.
+    bytes Password = 3;
+    // U2FRegisterResponse is U2F registration challenge response.
+    U2FRegisterResponse U2FRegisterResponse = 4;
+    // DeviceName is the name of the second factor device.
+    string DeviceName = 5;
+}
+
+// ChangePasswordWithTokenResponse is a response for ChangePasswordWithToken.
+message ChangePasswordWithTokenResponse {
+    // WebSession is a user's web sesssion created from successful changing of password.
+    types.WebSessionV2 WebSession = 1;
+    // RecoveryCodes are user's new recovery codes.
+    repeated string RecoveryCodes = 2;
+}
+
+// CreateRecoveryStartTokenRequest defines a request to create a user token for starting account
+// recovery.
+message CreateRecoveryStartTokenRequest {
+    // Username is the requesting user.
+    string Username = 1;
+    // RecoveryCode is one of the user's recovery code.
+    bytes RecoveryCode = 2;
+    // IsRecoverPassword is a flag to indicate if request is to recover password or second factor.
+    bool IsRecoverPassword = 3;
+}
+
+// AuthenticateUserWithRecoveryTokenRequest defines a request to authenticate the user defined in
+// token with either password or a second factor.
+message AuthenticateUserWithRecoveryTokenRequest {
+    // TokenID is the ID of a user recovery token.
+    string TokenID = 1;
+    // Username is the name of the user that the token belongs to.
+    string Username = 2;
+    // AuthCred is either a user's password or one of their second factor.
+    oneof AuthCred {
+        // SecondFactorToken is 2nd factor token value.
+        string SecondFactorToken = 3;
+        // Password is user password.
+        bytes Password = 4;
+        // U2FSignResponse is U2F sign response to a challenge.
+        U2FResponse U2FSignResponse = 5;
+    }
+}
+
+// SetNewAuthCredWithRecoveryTokenRequest defines a request to set either a new password or
+// add a new mfa device.
+message SetNewAuthCredWithRecoveryTokenRequest {
+    // TokenID is the ID of a user recovery token.
+    string TokenID = 1;
+    // DeviceName is the name of a new mfa device.
+    string DeviceName = 2;
+    // AuthCred is either a user's password or one of their second factor.
+    oneof AuthCred {
+        // SecondFactorToken is 2nd factor token value.
+        string SecondFactorToken = 3;
+        // Password is user password.
+        bytes Password = 4;
+        // U2FRegisterResponse is U2F registration challenge response.
+        U2FRegisterResponse U2FRegisterResponse = 5;
+    }
+}
+
+// CreateRecoveryCodesWithTokenRequest is a request to create new recovery codes for the user
+// defined in token.
+message CreateRecoveryCodesWithTokenRequest {
+    // TokenID is the ID of a user recovery token.
+    string TokenID = 1;
```

Comment:
I'm not sure whether this should contain MFAAuthenticateResponse instead of just a token.

Reply:
MFA is not required here, just a token to protect this request b/c this request can be made without a user being logged in.